### ISSUE TITLE
Deference of possible null reference after checking and returning

### ISSUE THREAD

**Issue by Axemasta** [dotnet/roslyn#82331]
**Version Used**: net10.0, 10.0.102

I noticed this warning being present when working with [`IMultiValueConverter`](https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/IMultiValueConverter.cs) in .NET Maui where they have the interface contract:
```csharp
#nullable disable
object Convert(object[] values, Type targetType, object parameter, CultureInfo culture);
```

Because they're supressing nullability, I have to check for nulls in the method implementation as the whole array can be null or elements in the array can be null, I've done so by marking the values array: `object?[]?`

**Steps to Reproduce**:

The following method produces a warning:
```csharp
string ConvertWarning(object?[]? values)
{
    if (values is null || values.Length != 3)
    {
        return string.Empty;
    }
    
    if (values[0] is null)
    {
        return string.Empty;
    }

    if (values[0] is not string)
    {
        var type = values[0].GetType().Name;
    }

    return string.Empty;
}
```

Warning:
```
0>Program.cs(21,20): Warning CS8602 : Dereference of a possibly null reference.
```

<img width="620" height="470" alt="Image" src="https://github.com/user-attachments/assets/fce1b3de-719e-40d1-8c89-196ce9ee8c70" />

If you store `values[0]` off as a local variable, you will not see the warning:

<img width="445" height="526" alt="Image" src="https://github.com/user-attachments/assets/a3ee731f-9b17-4840-ab2e-1adbe974075c" />

**Expected Behavior**:

We have checked for null already, the compiler should be satisfied the value is not null

**Actual Behavior**:

The warning `Warning CS8602 : Dereference of a possibly null reference.` is generated when accessing the element in the array we've already checked.

> This is alleviated by using the null forgiving operator which is fine for me but just behaviour I didn't expect
> `var type = values[0]!.GetType().Name;`

**Comment by MihuBot** [comment]
I'm a bot. Here are possible duplicate issues (I may be wrong):
- https://github.com/dotnet/roslyn/issues/34644
- https://github.com/dotnet/roslyn/issues/49653
- https://github.com/dotnet/roslyn/issues/33278
- https://github.com/dotnet/roslyn/issues/29964

**Comment by jjonescz** [comment]
Thanks for taking the time to file this issue. The current behavior looks to be by design, see the linked issues. cc @RikkiGibson
